src/material-experimental/theming/_theming.scss

```diff
@@ -0,0 +1,142 @@
+@use 'sass:list';
+@use 'sass:map';
+@use 'sass:meta';
+@use 'sass:string';
+@use '@angular/material' as mat;
+
+// Whether to throw an error when a required dep is not configured. If false, the dep will be
+// automatically configured instead.
+$_error-on-missing-dep: false;
+
+// Applies the theme for the given component configuration.
+@mixin _apply-theme($tokens, $component) {
+  $id: map.get($component, id);
+  $tokens: map.deep-merge($tokens, map.get($component, customizations));
+
+  // NOTE: for now we use a hardcoded if-chain, but in the future when first-class mixins are
+  // supported, the configuration data will contain a reference to its own theme mixin.
+  @if $id == 'mat.card' {
+    @include mat.card-theme($tokens);
+  } @else if $id == 'mat.checkbox' {
+    @include mat.checkbox-theme($tokens);
+  } @else {
+    @error 'Unrecognized component theme: #{id}';
+  }
+}
+
+// Gets the transitive dependency configurations for the given list of component configurations.
+@function _get-transitive-deps($components, $configured: ()) {
+  // Mark the given components as configured.
+  @each $component in $components {
+    $configured: map.set($configured, map.get($component, id), true);
+  }
+  $new-deps: ();
+
+  // Check each of the given components for new deps.
+  @each $component in $components {
+    @each $dep-getter in mat.private-normalize-args-list(map.get($component, deps)) {
+      $dep: meta.call($dep-getter);
+      $dep-id: map.get($dep, id);
+      @if not (map.has-key($configured, $dep-id)) {
```

Comment:
Should there also be an error somewhere if a given `id` has already been registered?

Reply:
I'm not sure. There's a few ways we could handle it:
1. Each component can only be configured once
2. Each component can be configured multiple times, as long as they're not setting the same token. e.g. there could be one config for mat.checkbox that sets label-color and a separate config that sets label-font-size and we just merge the maps to get the full set of token customizations
3. Each component can be configured multiple times including setting the same token, we just have well defined rules for which will win. In the current implementation, the one later in the list will win